src/Evaluator.hs

```diff
@@ -0,0 +1,214 @@
+    return $ TupleVal valList
+evalTuple expr _ = exprTypeError expr "Tuple"
+
+evalArray :: Show a => Expr a -> Environment -> IO Value
+evalArray (Array exprList _ _) env = do
+    valList <- evalExprList exprList env
+    return $ TupleVal valList
+evalArray expr _ = exprTypeError expr "Array"
+
+matchPatternList :: [Pattern a] -> [Value] -> Maybe Environment
+matchPatternList [] [] = Just emptyEnvironment
+matchPatternList [] _ = Nothing
+matchPatternList _ [] = Nothing
+matchPatternList (pattern : restPatterns) (val : restVals) = do
+    firstMatch <- matchPattern pattern val
+    restMatch  <- matchPatternList restPatterns restVals
+    return (unionEnvironments firstMatch restMatch)
+
+matchPattern :: Pattern a -> Value -> Maybe Environment
+matchPattern (WildcardPattern _ _) _ = Just emptyEnvironment
+matchPattern (ConstPattern primExpr _ _) (PrimitiveVal primVal)
+    | (evalPrimExpr primExpr) == primVal = Just emptyEnvironment
+    | otherwise                          = Nothing
+matchPattern ConstPattern{} _ = Nothing
+matchPattern (TuplePattern patternList _ _) (TupleVal valList)
+    | (length patternList) == (length valList) =
+        matchPatternList patternList valList
+    | otherwise = Nothing
+matchPattern TuplePattern{} _ = Nothing
+matchPattern (VarPattern identifier _ _) val =
+        Just $ insertIdentifier identifier val emptyEnvironment
+matchPattern (NamedPattern identifier pattern _ _) val = do
+    specificMatch <- matchPattern pattern val
+    return $ insertIdentifier identifier val specificMatch
+matchPattern (ConstructorPattern patternConstructorName patterns _ _)
+            (CustomType valConstructorName values)
+    | patternConstructorName == valConstructorName = matchPatternList patterns values
+    | otherwise = Nothing
+matchPattern (ConstructorPattern {}) _ = Nothing
+
+apply :: Value -> Value -> IO Value
+apply (Procedure pattern body procEnv) argument = 
+    case matchPattern pattern argument of
+        Just applyEnv -> eval body (unionEnvironments procEnv applyEnv)
+-- TODO: apply to something else?
+
+evalApp :: Show a => Expr a -> Environment -> IO Value
+evalApp (App operator argument _) env = do
+    opVal  <- eval operator env
+    argVal <- eval argument env
+    apply opVal argVal
+evalApp expr _ = exprTypeError expr "App"
+
+evalMatch :: Show a => Expr a -> Environment -> IO Value
+evalMatch (Match expr matchCases _ _) env =  do
+    value <- eval expr env
+    firstMatch value matchCases env where
+        firstMatch _ [] env = error "No match"
+        firstMatch val ((MatchCase pattern body):nextCases) env =
+            case matchPattern pattern val of
+                (Just applyEnv) -> eval body (unionEnvironments applyEnv env)
+                Nothing -> firstMatch val nextCases env
+
+evalExternal :: Show a => Expr a -> Environment -> IO Value
+evalExternal (External filePath name _ _) env = undefined
+-- TODO: Parse external file and find identifier value?
```

Comment:
This should somehow dynamically load a ".dll" or ".so" library or something like that. I don't think it should be implemented for the evaluator though.